tools/lambda-promtail/template.yaml

```diff
@@ -19,43 +15,72 @@ Parameters:
     Default: 2
 
 Resources:
+  LambdaPromtailRole:
+    Type: AWS::IAM::Role
+    Properties:
+      AssumeRolePolicyDocument:
+        Version: '2012-10-17'
+        Statement:
+        - Effect: Allow
+          Principal:
+            Service:
+            - lambda.amazonaws.com
+          Action:
+          - sts:AssumeRole
+      Description: "Lambda Promtail Role"
+      Policies:
+      - PolicyName: logs
+        PolicyDocument:
+          Version: '2012-10-17'
+          Statement:
+          - Effect: Allow
+            Action:
+            - logs:CreateLogGroup
+            - logs:CreateLogStream
+            - logs:PutLogEvents
+            Resource: arn:aws:logs:*:*:*
+      RoleName: iam_for_lambda
   LambdaPromtailFunction:
-    Type: AWS::Serverless::Function
+    Type: AWS::Lambda::Function
     Properties:
-      CodeUri: lambda-promtail/
-      Handler: handler
+      Code:
+        ImageUri: "your-image-ecr:latest"
       MemorySize: 128
+      PackageType: Image
       Timeout: 60
-      EventInvokeConfig:
-        MaximumRetryAttempts: 2
-      Policies:
-        - Statement:
-            Effect: Allow
-            Action:
-              - 'logs:DescribeLogGroups'
-              - 'logs:DescribeLogStreams'
-              - 'logs:FilterLogEvents'
-              - 'logs:GetLogEvents'
-            Resource: '*'
+      Role: !GetAtt LambdaPromtailRole.Arn
       ReservedConcurrentExecutions: !Ref ReservedConcurrency
       # # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-lambda-function-vpcconfig.html
       # VpcConfig:
-      Runtime: go1.x
-      Events:
-        CatchAll:
-          Type: CloudWatchLogs
-          Properties:
-            LogGroupName: !Ref LogGroup
-            FilterPattern: ''
       Environment: # More info about Env Vars: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#environment-object
         Variables:
-          PROMTAIL_ADDRESS: !Ref PromtailAddress
-
+          WRITE_ADDRESS: !Ref WriteAddress
+  LambdaPromtailVersion:
+    Type: AWS::Lambda::Version
+    Properties:
+      FunctionName: !Ref LambdaPromtailFunction
+  LambdaPromtailPermissions:
+    Type: AWS::Lambda::Permission
+    Properties:
+      Action: lambda:InvokeFunction
+      FunctionName: !GetAtt LambdaPromtailFunction.Arn
+      Principal: !Sub
+        - logs.${Region}.amazonaws.com
+        - { Region: !Ref "AWS::Region" }
+  LambdaPromtailEventInvokeConfig:
+    Type: AWS::Lambda::EventInvokeConfig
+    Properties:
+      FunctionName: !Ref LambdaPromtailFunction
+      MaximumRetryAttempts: 2
+      Qualifier: !GetAtt LambdaPromtailVersion.Version
+  MainLambdaPromtailSubscriptionFilter:
+    Type: AWS::Logs::SubscriptionFilter
+    Properties:
+      DestinationArn: !GetAtt LambdaPromtailFunction.Arn
+      FilterPattern: ""
+      LogGroupName: "/aws/lambda/some-lamda-log-group"
```

Comment:
Again, let's use parameters.

Reply:
This one I'd prefer to leave as is, since it's more likely than not that users will have multiple log groups that they want to subscribe to, with different `FilterPattern`s per group. 

However, I'll have a look at if there's a nice way to duplicate a terraform config block with an array as an input.

Reply:
Unfortunately there isn't a nice built in way to repeat a resource block in CloudFormation based on a list of variables, but I have made that change to the Terraform file. IMO it's best if users copy paste this section for CloudFormation until we find a better option here.